x-pack/plugins/security/server/lib/authentication/authenticator.js

```diff
@@ -156,9 +148,12 @@ class Authenticator {
         const sessionCanBeUpdated = (authenticationResult.succeeded() || authenticationResult.redirected())
           && (authenticationResult.state || !isSystemApiRequest);
 
-        // If provider owned the session, but failed to authenticate anyway, that likely means
-        // that session is not valid and we should clear it.
-        if (authenticationResult.failed() && getErrorStatusCode(authenticationResult.error) === 401) {
+        // If provider owned the session, but failed to authenticate anyway, that likely means that
+        // session is not valid and we should clear it. Also provider can specifically ask to clear
+        // session by setting it to `null` even if authentication attempt didn't fail.
+        if (authenticationResult.state === null || (
```

Comment:
__note:__ alternatively (instead of having different meaning for `null` and `undefined`) we could use something like this:
```ts
class AuthenticationResult {
  static ClearSession = Symbol('clear session');
  .....
}
.....
AuthenticationResult.redirect(....., AuthenticationResult.ClearSession);
```

IMO it'd make sense if we had several different symbols, but it's the only use case I can think of and using `null` feels okay to me. What do you think?

Reply:
Good idea! Will do.